src/types.ts

```diff
@@ -219,22 +225,13 @@ export namespace OfCoerce {
         true extends Utils.IsUnion<Type>
         ? // Resolve union
           // First add the defined coercer (String, Number, etc.)
-          | Core.Union<
-                Type extends SchemaPair<any, infer Coercer> ? Coercer : never
-              >
-            // Now add the custom coercer
-            | (Flag extends "root"
-                ? never
-                : Type extends SchemaPair<infer Type, any>
-                ? Core.Coercer<Type>
-                : never)
+          | Core.Union<(Type extends SchemaPair<infer Type, infer Coercer> ? Coercer extends never ? Core.Coercer<Type> : Coercer :  never)[]>
         : Type extends SchemaPair<infer Type, infer Coercer>
         ?
             | // I wrap it into the union at the end to make it chance to resolve to
             // the final type to avoid false positive on the union check.
-            (true extends Utils.IsUnion<Type> ? Core.Union<Coercer> : Coercer)
+            (true extends Utils.IsUnion<Type> ? Core.Union<Coercer[]> : Coercer)
             // Add custom coercer
-            | (Flag extends "root" ? never : Core.Coercer<Type>)
         : never
```

Comment:
This was either a bug or a design bug, for the type
```ts
interface Baz {
  value: Bar | Foo
}
```
the schema was
```ts
type Schema = {
  value: Union<({...},{...})[]> | Coerser<Foo> | Coerser<Bar>
}
```
The expected type should probably be
```ts
type Schema = {
  value: Union<(Coerser<Foo> | Coerser<Bar>, {...}, {...})[]>
}
```

Note that because of the initial design of `Union` that held the literal union in the type (instead of a tuple representing the union) it was too invasive to update the entire api for that... sorry for the partial work, but `infer` would be accurate while `Defined` would be accepting of more flex

Reply:
I definitely lost my battle with this type, thank you for fixing the bug!